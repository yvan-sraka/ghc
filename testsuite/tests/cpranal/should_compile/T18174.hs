{-# OPTIONS_GHC -O2 -fforce-recomp #-}
{-# LANGUAGE BangPatterns #-}

module T18174 (fac1, fac2, fac3, facIO, h1, h2) where

----------------------------------------------------------------------
-- First some basic examples that we want to CPR nestedly.

-- pretty strict
fac1 :: Int -> a ->  (a, Int)
fac1 n s | n < 2     = (s,1)
         | otherwise = case  fac1 (n-1) s of (s',n') -> let n'' = n*n' in n'' `seq` (s',n'')

-- lazier, but Int still has CPR
fac2 :: Int -> a ->  (a, Int)
fac2 n s | n < 2     = (s,1)
         | otherwise = case  fac2 (n-1) s of (s',n') -> (s',n'*n')

-- even lazier, but evaluation of the Int doesn't terminate rapidly!
-- Thus, we may not WW for the nested Int.
-- Otherwise @fac3 99999 () `seq` ()@ (which should terminate rapidly)
-- evaluates more than necessary.
fac3 :: Int -> a ->  (a, Int)
fac3 n s | n < 2     = (s,1)
         | otherwise = let (s',n') = fac3 (n-1) s in (s',n'*n')

facIO :: Int -> IO Int
facIO n | n < 2     = return 1
        | otherwise = do n' <- facIO (n-1); return (n*n')

----------------------------------------------------------------------
-- The following functions are copied from T18894. This test is about
-- *exploiting* the demand signatures that we assertedly (by T18894)
-- annotate.

g1 :: Int -> (Int,Int)
g1 1 = (15, 0)
g1 n = (2 * n, 2 `div` n)
{-# NOINLINE g1 #-}

-- | Sadly, the @g1 2@ subexpression will be floated to top-level, where we
-- don't see the specific demand placed on it by @snd@. Tracked in #19001.
h1 :: Int -> Int
h1 1 = 0
h1 2 = snd (g1 2)
h1 m = uncurry (+) (g1 m)

-- | So @g2@ here takes an additional argument m that prohibits floating to
-- top-level. We want that argument to have the CPR property, so we have
-- to add a bang so that it's used strictly and ultimately unboxed.
-- We expect the following CPR type:
--
-- > #c1(#c1(#), *c1(#))
--
-- In combination with the the fact that all calls to @g2@ evaluate the second
-- component of the pair, we may unbox @g2@ to @(# Int#, Int# #)@.
g2 :: Int -> Int -> (Int,Int)
g2 !m 1 = (2 + m, 0)
g2 m  n = (2 * m, 2 `div` n)
{-# NOINLINE g2 #-}

h2 :: Int -> Int
h2 1 = 0
h2 m
  | odd m     = snd (g2 m 2)
  | otherwise = uncurry (+) (g2 2 m)
