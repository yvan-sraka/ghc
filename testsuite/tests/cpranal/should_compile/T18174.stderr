
==================== Tidy Core ====================
Result size of Tidy Core = {terms: 336, types: 368, coercions: 6, joins: 0/1}

-- RHS size: {terms: 1, types: 0, coercions: 0, joins: 0/0}
T18174.$trModule4 :: GHC.Prim.Addr#
T18174.$trModule4 = "main"#

-- RHS size: {terms: 2, types: 0, coercions: 0, joins: 0/0}
T18174.$trModule3 :: GHC.Types.TrName
T18174.$trModule3 = GHC.Types.TrNameS T18174.$trModule4

-- RHS size: {terms: 1, types: 0, coercions: 0, joins: 0/0}
T18174.$trModule2 :: GHC.Prim.Addr#
T18174.$trModule2 = "T18174"#

-- RHS size: {terms: 2, types: 0, coercions: 0, joins: 0/0}
T18174.$trModule1 :: GHC.Types.TrName
T18174.$trModule1 = GHC.Types.TrNameS T18174.$trModule2

-- RHS size: {terms: 3, types: 0, coercions: 0, joins: 0/0}
T18174.$trModule :: GHC.Types.Module
T18174.$trModule = GHC.Types.Module T18174.$trModule3 T18174.$trModule1

-- RHS size: {terms: 2, types: 0, coercions: 0, joins: 0/0}
lvl_r1mk :: Int
lvl_r1mk = GHC.Types.I# 1#

Rec {
-- RHS size: {terms: 38, types: 38, coercions: 0, joins: 0/1}
T18174.$wfac3 :: forall {a}. GHC.Prim.Int# -> a -> (# a, Int #)
T18174.$wfac3
  = \ (@a_s1d4) (ww_s1d9 :: GHC.Prim.Int#) (w_s1d6 :: a_s1d4) ->
      case GHC.Prim.<# ww_s1d9 2# of {
        __DEFAULT ->
          let {
            ds_s186 :: (a_s1d4, Int)
            ds_s186 = case T18174.$wfac3 @a_s1d4 (GHC.Prim.-# ww_s1d9 1#) w_s1d6 of { (# ww2_s1dd, ww3_s1de #) -> (ww2_s1dd, ww3_s1de) } } in
          (# case ds_s186 of { (s'_aXb, n'_aXc) -> s'_aXb }, case ds_s186 of { (s'_aXb, n'_aXc) -> case n'_aXc of { GHC.Types.I# ww2_s1d2 -> GHC.Types.I# (GHC.Prim.*# ww2_s1d2 ww2_s1d2) } } #);
        1# -> (# w_s1d6, lvl_r1mk #)
      }
end Rec }

-- RHS size: {terms: 14, types: 16, coercions: 0, joins: 0/0}
fac3 :: forall a. Int -> a -> (a, Int)
fac3 = \ (@a_s1d4) (w_s1d5 :: Int) (w1_s1d6 :: a_s1d4) -> case w_s1d5 of { GHC.Types.I# ww1_s1d9 -> case T18174.$wfac3 @a_s1d4 ww1_s1d9 w1_s1d6 of { (# ww3_s1dd, ww4_s1de #) -> (ww3_s1dd, ww4_s1de) } }

Rec {
-- RHS size: {terms: 24, types: 21, coercions: 0, joins: 0/0}
T18174.$wfac2 :: forall {a}. GHC.Prim.Int# -> a -> (# a, GHC.Prim.Int# #)
T18174.$wfac2
  = \ (@a_s1dh) (ww_s1dm :: GHC.Prim.Int#) (w_s1dj :: a_s1dh) ->
      case GHC.Prim.<# ww_s1dm 2# of {
        __DEFAULT -> case T18174.$wfac2 @a_s1dh (GHC.Prim.-# ww_s1dm 1#) w_s1dj of { (# ww2_s1ds, ww3_s1du #) -> (# ww2_s1ds, GHC.Prim.*# ww3_s1du ww3_s1du #) };
        1# -> (# w_s1dj, 1# #)
      }
end Rec }

-- RHS size: {terms: 15, types: 16, coercions: 0, joins: 0/0}
fac2 :: forall a. Int -> a -> (a, Int)
fac2 = \ (@a_s1dh) (w_s1di :: Int) (w1_s1dj :: a_s1dh) -> case w_s1di of { GHC.Types.I# ww1_s1dm -> case T18174.$wfac2 @a_s1dh ww1_s1dm w1_s1dj of { (# ww3_s1ds, ww4_s1du #) -> (ww3_s1ds, GHC.Types.I# ww4_s1du) } }

Rec {
-- RHS size: {terms: 24, types: 21, coercions: 0, joins: 0/0}
T18174.$wfac1 :: forall {a}. GHC.Prim.Int# -> a -> (# a, GHC.Prim.Int# #)
T18174.$wfac1
  = \ (@a_s1dx) (ww_s1dC :: GHC.Prim.Int#) (w_s1dz :: a_s1dx) ->
      case GHC.Prim.<# ww_s1dC 2# of {
        __DEFAULT -> case T18174.$wfac1 @a_s1dx (GHC.Prim.-# ww_s1dC 1#) w_s1dz of { (# ww2_s1dI, ww3_s1dK #) -> (# ww2_s1dI, GHC.Prim.*# ww_s1dC ww3_s1dK #) };
        1# -> (# w_s1dz, 1# #)
      }
end Rec }

-- RHS size: {terms: 15, types: 16, coercions: 0, joins: 0/0}
fac1 :: forall a. Int -> a -> (a, Int)
fac1 = \ (@a_s1dx) (w_s1dy :: Int) (w1_s1dz :: a_s1dx) -> case w_s1dy of { GHC.Types.I# ww1_s1dC -> case T18174.$wfac1 @a_s1dx ww1_s1dC w1_s1dz of { (# ww3_s1dI, ww4_s1dK #) -> (ww3_s1dI, GHC.Types.I# ww4_s1dK) } }

-- RHS size: {terms: 30, types: 19, coercions: 0, joins: 0/0}
T18174.$wg2 :: GHC.Prim.Int# -> GHC.Prim.Int# -> (# GHC.Prim.Int#, GHC.Prim.Int# #)
T18174.$wg2
  = \ (ww_s1dR :: GHC.Prim.Int#) (ww1_s1dV :: GHC.Prim.Int#) ->
      case ww1_s1dV of ds_X2 {
        __DEFAULT -> case GHC.Classes.divInt# 2# ds_X2 of ww4_a153 { __DEFAULT -> (# GHC.Prim.*# 2# ww_s1dR, ww4_a153 #) };
        -1# -> (# GHC.Prim.*# 2# ww_s1dR, -2# #);
        0# -> case GHC.Real.divZeroError of wild_00 { };
        1# -> (# GHC.Prim.+# 2# ww_s1dR, 0# #)
      }

-- RHS size: {terms: 26, types: 17, coercions: 0, joins: 0/0}
T18174.$wh2 :: GHC.Prim.Int# -> GHC.Prim.Int#
T18174.$wh2
  = \ (ww_s1ed :: GHC.Prim.Int#) ->
      case ww_s1ed of ds_X2 {
        __DEFAULT ->
          case GHC.Prim.remInt# ds_X2 2# of {
            __DEFAULT -> case T18174.$wg2 ds_X2 2# of { (# ww2_s1e4, ww3_s1e7 #) -> ww3_s1e7 };
            0# -> case T18174.$wg2 2# ds_X2 of { (# ww2_s1e4, ww3_s1e7 #) -> GHC.Prim.+# ww2_s1e4 ww3_s1e7 }
          };
        1# -> 0#
      }

-- RHS size: {terms: 10, types: 4, coercions: 0, joins: 0/0}
h2 :: Int -> Int
h2 = \ (w_s1ea :: Int) -> case w_s1ea of { GHC.Types.I# ww1_s1ed -> case T18174.$wh2 ww1_s1ed of ww2_s1eh { __DEFAULT -> GHC.Types.I# ww2_s1eh } }

-- RHS size: {terms: 2, types: 0, coercions: 0, joins: 0/0}
T18174.h5 :: Int
T18174.h5 = GHC.Types.I# 0#

-- RHS size: {terms: 2, types: 0, coercions: 0, joins: 0/0}
lvl1_r1ml :: Int
lvl1_r1ml = GHC.Types.I# -2#

-- RHS size: {terms: 27, types: 15, coercions: 0, joins: 0/0}
T18174.$wg1 :: GHC.Prim.Int# -> (# GHC.Prim.Int#, Int #)
T18174.$wg1
  = \ (ww_s1en :: GHC.Prim.Int#) ->
      case ww_s1en of ds_X2 {
        __DEFAULT ->
          (# GHC.Prim.*# 2# ds_X2,
             case ds_X2 of {
               __DEFAULT -> case GHC.Classes.divInt# 2# ds_X2 of ww4_a153 { __DEFAULT -> GHC.Types.I# ww4_a153 };
               -1# -> lvl1_r1ml;
               0# -> case GHC.Real.divZeroError of wild1_00 { }
             } #);
        1# -> (# 15#, T18174.h5 #)
      }

-- RHS size: {terms: 8, types: 9, coercions: 0, joins: 0/0}
T18174.h4 :: (Int, Int)
T18174.h4 = case T18174.$wg1 2# of { (# ww1_s1eu, ww2_s1ew #) -> (GHC.Types.I# ww1_s1eu, ww2_s1ew) }

-- RHS size: {terms: 22, types: 16, coercions: 0, joins: 0/0}
T18174.$wh1 :: GHC.Prim.Int# -> Int
T18174.$wh1
  = \ (ww_s1eC :: GHC.Prim.Int#) ->
      case ww_s1eC of ds_X2 {
        __DEFAULT -> case T18174.$wg1 ds_X2 of { (# ww2_s1eu, ww3_s1ew #) -> case ww3_s1ew of { GHC.Types.I# y_a15e -> GHC.Types.I# (GHC.Prim.+# ww2_s1eu y_a15e) } };
        1# -> T18174.h5;
        2# -> case T18174.h4 of { (ds1_a137, y_a138) -> y_a138 }
      }

-- RHS size: {terms: 6, types: 3, coercions: 0, joins: 0/0}
h1 :: Int -> Int
h1 = \ (w_s1ez :: Int) -> case w_s1ez of { GHC.Types.I# ww1_s1eC -> T18174.$wh1 ww1_s1eC }

Rec {
-- RHS size: {terms: 23, types: 29, coercions: 0, joins: 0/0}
T18174.$wfacIO :: GHC.Prim.Int# -> GHC.Prim.State# GHC.Prim.RealWorld -> (# GHC.Prim.State# GHC.Prim.RealWorld, GHC.Prim.Int# #)
T18174.$wfacIO
  = \ (ww_s1eJ :: GHC.Prim.Int#) (w_s1eG :: GHC.Prim.State# GHC.Prim.RealWorld) ->
      case GHC.Prim.<# ww_s1eJ 2# of {
        __DEFAULT -> case T18174.$wfacIO (GHC.Prim.-# ww_s1eJ 1#) w_s1eG of { (# ww2_s1eP, ww3_s1eR #) -> (# ww2_s1eP, GHC.Prim.*# ww_s1eJ ww3_s1eR #) };
        1# -> (# w_s1eG, 1# #)
      }
end Rec }

-- RHS size: {terms: 14, types: 23, coercions: 0, joins: 0/0}
T18174.facIO1 :: Int -> GHC.Prim.State# GHC.Prim.RealWorld -> (# GHC.Prim.State# GHC.Prim.RealWorld, Int #)
T18174.facIO1 = \ (w_s1eF :: Int) (w1_s1eG :: GHC.Prim.State# GHC.Prim.RealWorld) -> case w_s1eF of { GHC.Types.I# ww1_s1eJ -> case T18174.$wfacIO ww1_s1eJ w1_s1eG of { (# ww3_s1eP, ww4_s1eR #) -> (# ww3_s1eP, GHC.Types.I# ww4_s1eR #) } }

-- RHS size: {terms: 1, types: 0, coercions: 6, joins: 0/0}
facIO :: Int -> IO Int
facIO = T18174.facIO1 `cast` (<Int>_R %<'Many>_N ->_R Sym (GHC.Types.N:IO[0] <Int>_R) :: (Int -> GHC.Prim.State# GHC.Prim.RealWorld -> (# GHC.Prim.State# GHC.Prim.RealWorld, Int #)) ~R# (Int -> IO Int))



